skoolkit/basic.py

```diff
@@ -194,24 +233,105 @@ def _get_num_str(self, j):
             num_str = chr(self.snapshot[j]) + num_str
         return num_str[1:]
 
-    def _get_char(self, code):
-        if code == 94:
-            return '↑'
-        if code == 96:
-            return '£'
-        if code == 127:
-            return '©'
-        if 32 <= code <= 126:
-            return chr(code)
-        return '{{0x{:02X}}}'.format(code)
+class VariableLister:
+    def __init__(self):
+        self.text = TextReader()
 
-    def _get_token(self, code):
-        token = TOKENS[code]
-        if self.lspace and code >= 197 and token[0] >= 'A':
-            token = ' ' + token
-        self.lspace = True
-        if code < 168 or code == 203 or token[-1] in '#=>':
-            # RND, INKEY$, PI, THEN, '<=', '>=', '<>', 'OPEN #', 'CLOSE #'
-            return token
-        self.lspace = False
-        return token + ' '
+    def list_variables(self, snapshot):
+        lines = []
+        self.snapshot = snapshot
+        i = snapshot[23627] + 256 * snapshot[23628]
+        while i < len(snapshot) and snapshot[i] != 128:
+            self.text.lspace = False
+            variable_type = snapshot[i] & 224
+            if variable_type == 64:
+                # String (010xxxxx)
+                i, line = self._get_string_var(i)
+            elif variable_type == 128:
+                # Array of numbers (100xxxxx)
+                i, line = self._get_num_array_var(i)
+            elif variable_type == 160:
+                # Number whose name is longer than one letter (101xxxxx)
+                i, line = self._get_long_num_var(i)
+            elif variable_type == 192:
+                # Array of characters (110xxxxx)
+                i, line = self._get_char_array_var(i)
+            elif variable_type == 224:
+                # Control variable of a FOR-NEXT loop (111xxxxx)
+                i, line = self._get_control_var(i)
+            elif variable_type == 96:
+                # Number whose name is one letter (011xxxxx)
+                i, line = self._get_short_num_var(i)
+            else:
+                # Basic line (000xxxxx / 001xxxxx)
+                i += get_word(snapshot, i + 2) + 4
+                continue
+            lines.append('{}'.format(line))
+        return '\n'.join(lines)
+
+    def _get_string_var(self, i):
+        letter = self.text.get_chars((self.snapshot[i] & 31) + 96)
+        str_length = get_word(self.snapshot, i + 1)
+        value = ''.join([self.text.get_chars(c) for c in self.snapshot[i + 3:i + 3 + str_length]])
+        line = '(String) {}$="{}"'.format(letter, value)
+        return i + 3 + str_length, line
+
+    def _get_num_array_var(self, i):
```

Comment:
This method would be clearer if each component of ``line`` were assigned to a suitably named variable, and then ``line`` were built from a single format string. For example:

    ...
    varname = self.text.get_chars(letter)
    ...
    dims = [str(get_word(self.snapshot, c)) for c in range(i, i + 2 * dimensions, 2)]
    ...
    values = [str(get_number(self.snapshot, c)) for c in range(i, i + data_length, 5)]
    ...
    line = '(Number array) {}({})=[{}]'.format(varname, ','.join(dims), ','.join(values))

Use the same approach with ``_get_long_num_var()`` and ``_get_char_array_var()`` too.